[FLINK-4355] [cluster management] Implement TaskManager side of registration at ResourceManager.

StephanEwen opened the pull request:
This implements the first part of the `TaskManager` (here temporarily called `TaskExecutor` to avoid name clashes) as designed in FLIP-6.

Specifically, it accepts a leader retrieval service for the `ResourceManager` and triggers the registration at the ResourceManager.

Tests are pending, this pull request is intended for early review.

**UPDATE:** I rebased this onto some other pull requests (which form necessary utilities).
Please view only https://github.com/apache/flink/pull/2353/commits/6e5059d6faab460012989fed7ff9ac62f42e3ffd and https://github.com/apache/flink/pull/2353/commits/77c54dfd7f4c50465c629e88995cb865318e8605 as changes of this pull request.

wenlong88 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutorToResourceManagerConnection.java:
```diff
@@ -0,0 +1,80 @@
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+public class TaskExecutorToResourceManagerConnection {
```
I think we can have a HARPCGateway extends HAService which can be reused for other components

wenlong88 replied:
I think we can have a HARPCGateway extends HAService which can be reused for other components

StephanEwen replied:
Agreed, let's factor it out as a common utility do that as part of the next component that needs it.

wenlong88 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutorToResourceManagerConnection.java:
```diff
@@ -0,0 +1,80 @@
+
+	private final String resourceManagerAddress;
+
+	public TaskExecutorToResourceManagerConnection(
```
I think we can have a HARPCGateway extends HAService which can be reused for other components

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/TaskExecutorRegistrationResponse.java:
```diff
@@ -0,0 +1,86 @@
+
+
+
+
```
Many line breaks.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/ResourceManagerRegistration.java:
```diff
@@ -0,0 +1,252 @@
+
+		Future<TaskExecutorRegistrationResponse> registrationFuture = resourceManager.registerTaskExecutor(
+				resourceManagerLeaderId, taskExecutor.getAddress(), taskExecutor.getResourceID(),
+				taskExecutor.getCurrentSlotReport(), timeout);
```
Do we have to get the current slot report for each registration attempt? Isn't this something which can be out of date anyway when it arrives at the RM? The heartbeat should consolidate the state in such a case, shouldn't it?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutorToResourceManagerConnection.java:
```diff
@@ -0,0 +1,132 @@
+
+
+	public TaskExecutorToResourceManagerConnection(
+			Logger log,
```
I'm not so sure whether passing a logger from a different context makes the logging statements easier or harder to understand. Usually I have troubles to locate the logging statements (and thus the state of the computation) when I see logging statements logged from a different (wider) scope than they actually happened.

tillrohrmann commented:
Really good code @StephanEwen :-) I like it a lot that you've pulled out the registration logic from the `TaskExecutor`. This makes it much cleaner and easier to test.

I'm actually wondering whether we can actually take it a step further. What I've noticed is that we have to pass the `TaskExecutor` to different places to have access to the its state and to run things in the main thread context. I'm not sure whether this is really necessary. 

The registration information should be more or less constant and the main thing which has to happen in the main thread context is the setting of the connected gateway. So wouldn't it be possible that the `ResourceManagerRegistration` is something like a cancelable future which is instantiated with all necessary information. It performs the registration logic completely in the background and once it is completed we can create a `TaskExecutorToResourceManagerConnection` in the main thread context. Whenever we have to start a new registration, we cancel the future (if there is still another registration running).

Pulling out the `TaskExecutor` from the registration process has the advantage that the components are even further separated and, thus, easier to test.

StephanEwen commented:
I like that idea.

Getting the TaskExecutor out of the registration means that we cannot transmit the slot report immediately with each registration call (it only makes sense when each attempt grabs the latest slot report).

That might not be too bad for the TaskExecutor, as it can always just eagerly transmit a slot report (such as when responding to a heartbeat).

StephanEwen commented:
I addressed the comments by @tillrohrmann and @wenlong88
  - turning the registration object into a reusable utility that can also be used for the JobMaster registration
  - I made the registration object behave more like a cancelable future. That way, it needs not use the TaskExecutor's main-thread-execution-context for its retries.

This looks actually pretty nice now. Will merge this to allow other people to build on top of it. Will provide tests ASAP.

StephanEwen commented:
Manually merged into the `flip-6` feature branch in https://github.com/apache/flink/commit/68addf39e0e5f9e1656818f923be362680ed93b0
